### PR TITLE
ZF3 compatibility (and general) improvements

### PR DESCRIPTION
**Pull request by zluiten**
#642 

 * Use EventManagerAwareTrait
 * Fixed resetAdapters test, removed deprecated/removed CallbackHandler
 * Type hint to UserInterface in db related classes (mapper, hydrator)
 * Type hint to interfaces in general where possible
 * Type hint to ContainerInterface instead of ServiceManager
 * Removed unused code
 * Code style clean up

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -25,31 +25,31 @@
...
-        "zendframework/zend-mvc-plugin-prg": "~1.0",
-        "zendframework/zend-hydrator": "~2.0"
+        "php": "^5.5|^7.0",
+        "zendframework/zend-authentication": "^2.5",
+        "zendframework/zend-crypt": "^3.0",
+        "zendframework/zend-form": "^2.9",
+        "zendframework/zend-inputfilter": "^2.7",
+        "zendframework/zend-loader": "^2.5",
+        "zendframework/zend-modulemanager": "^2.7",
+        "zendframework/zend-mvc": "^3.0",
```
Comment (snapshotpl) [inline composer.json:35]: Cannot you keep compatibility with v2 and v3?

**Comment by phgeek** [comment]
Hi @netiul I'm using your PR and got an error when login in:

```
File:
vendor/zf-commons/zfc-user/src/ZfcUser/Authentication/Adapter/AdapterChain.php:90
Message:
Call to a member function getListeners() on null
```

By commenting the _resetAdapters_ function content it works. Any idea on what's wrong?

PS: Using along with ZfcUserDoctrineORM.

**Comment by hilyafadhilah** [comment]
@phgeek @netiul I found out that _AdapterChain_ class calls `getEventManager()` that lazy-loads an _EventManager_ instance if none is registered. This lazy-loaded instance have `$sharedEventManager` defaults to null. This is the line in `resetAdapters()` that made the error.

    $listeners = $this->getEventManager()->getSharedManager()->getListeners(['authenticate'], 'authenticate');

Basically, `getSharedManager()` returned null because no `$sharedEventManager` was set. And so, the call for `getListeners()` trigerred an error. The fix would be to check first if it's null, right?

Also, I noticed that ZfcUserDoctrineORM should also be changed to be compatible with the PR. `ZfcUserDoctrineORM/Mapper/User::insert()` and `ZfcUserDoctrineORM/Mapper/User::update()` methods must use type hint now.

Other than that, this PR is really cool. 👍

**Comment by Danielss89** [comment]
Thank you, your changes have been introduced in 3.x!

**Comment by jroedel** [comment]
Great work Daniel! Thanks a lot for taking on this huge project. 
Users: We've been waiting a long time for this, let's work to upgrade to 3.0 to work out any bugs quickly. I'm sure I'm not the only one who's still on ZF2, waiting for ZfcUser to upgrade.

**Pull request closed without merging**